### ISSUE TITLE
Android: Support OIDC in REST

### ISSUE THREAD

**Issue by dnys1** [aws-amplify/amplify-flutter#978]
Ticket to track OIDC support in REST on Android.

**Comment by ObeA** [comment]
I'm trying to get a custom OIDC provider to work with the Amplify API plugin based on [this guide](https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/flutter/#oidc). On iOS it works, but on Android I keep getting the message "[REST OIDC/Lambda is not supported yet](https://github.com/aws-amplify/amplify-flutter/blob/75264e817a2d5aba97979443541106bacb475a6b/packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/auth/FlutterAuthProvider.kt#L73)" followed by the error below. Is this due to the lack of support of OIDC in Android (this ticket)? If necessary I can provide minimal repro.

```
E/amplify:flutter:api(23491): ApiException{message=OkHttp client failed to make a successful request., cause=ApiAuthException{message=Failed to retrieve auth token from OIDC provider., cause=ApiAuthException{message=Unable to retrieve token for OPENID_CONNECT, cause=null, recoverySuggestion=Make sure you register your auth providers in the addPlugin call and
E/amplify:flutter:api(23491): that getLatestAuthToken returns a value.}, recoverySuggestion=Check the application logs for details.}, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.AppSyncGraphQLOperation.start(AppSyncGraphQLOperation.java:113)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.AWSApiPlugin.mutate(AWSApiPlugin.java:265)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.AWSApiPlugin.mutate(AWSApiPlugin.java:252)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.ApiCategory.mutate(ApiCategory.java:70)
E/amplify:flutter:api(23491): 	at com.amazonaws.amplify.amplify_api.FlutterGraphQLApi$Companion.mutate(FlutterGraphQLApi.kt:173)
E/amplify:flutter:api(23491): 	at com.amazonaws.amplify.amplify_api.AmplifyApiPlugin.onMethodCall(AmplifyApiPlugin.kt:102)
E/amplify:flutter:api(23491): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/amplify:flutter:api(23491): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:84)
E/amplify:flutter:api(23491): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:865)
E/amplify:flutter:api(23491): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/amplify:flutter:api(23491): 	at android.os.MessageQueue.next(MessageQueue.java:335)
E/amplify:flutter:api(23491): 	at android.os.Looper.loop(Looper.java:183)
E/amplify:flutter:api(23491): 	at android.app.ActivityThread.main(ActivityThread.java:7656)
E/amplify:flutter:api(23491): 	at java.lang.reflect.Method.invoke(Native Method)
E/amplify:flutter:api(23491): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/amplify:flutter:api(23491): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E/amplify:flutter:api(23491): Caused by: ApiAuthException{message=Failed to retrieve auth token from OIDC provider., cause=ApiAuthException{message=Unable to retrieve token for OPENID_CONNECT, cause=null, recoverySuggestion=Make sure you register your auth providers in the addPlugin call and
E/amplify:flutter:api(23491): that getLatestAuthToken returns a value.}, recoverySuggestion=Check the application logs for details.}
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.auth.ApiRequestDecoratorFactory.forAuthType(ApiRequestDecoratorFactory.java:143)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.auth.ApiRequestDecoratorFactory.fromGraphQLRequest(ApiRequestDecoratorFactory.java:100)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.AppSyncGraphQLOperation.start(AppSyncGraphQLOperation.java:97)
E/amplify:flutter:api(23491): 	... 15 more
E/amplify:flutter:api(23491): Caused by: ApiAuthException{message=Unable to retrieve token for OPENID_CONNECT, cause=null, recoverySuggestion=Make sure you register your auth providers in the addPlugin call and
E/amplify:flutter:api(23491): that getLatestAuthToken returns a value.}
E/amplify:flutter:api(23491): 	at com.amazonaws.amplify.amplify_api.auth.FlutterAuthProvider$Companion.noTokenAvailable(FlutterAuthProvider.kt:139)
E/amplify:flutter:api(23491): 	at com.amazonaws.amplify.amplify_api.auth.FlutterAuthProvider.getLatestAuthToken(FlutterAuthProvider.kt:149)
E/amplify:flutter:api(23491): 	at com.amplifyframework.api.aws.auth.ApiRequestDecoratorFactory.forAuthType(ApiRequestDecoratorFactory.java:141)
E/amplify:flutter:api(23491): 	... 17 more
```

**Comment by dnys1** [comment]
Hi @ObeA, you've mentioned a link the GraphQL guide. GraphQL currently does support OIDC/Lambda on both platforms. However, as mentioned in the [REST](https://docs.amplify.aws/lib/restapi/authz/q/platform/flutter/#oidc) guide, Android does not support OIDC/Lambda in REST yet. This is a limitation in the Android framework we are currently working with the team to fix. Sorry for the inconvenience, and please stay tuned for a fix!

**Comment by ObeA** [comment]
Hi @dnys1, thank you for the quick response. Maybe I'm misunderstanding something, but as can be seen in the stacktrace I'm calling `mutate` on the GraphQL API. Should I open a new issue for the error I'm getting?

**Comment by dnys1** [comment]
Yes, @ObeA, please open a new issue with your logs and I will investigate.

**Comment by ObeA** [comment]
Opened #1022

**Comment by dnys1** [comment]
This issue has been fully resolved in 0.2.8
